nuxeo-core/nuxeo-core-storage-dbs/src/main/java/org/nuxeo/ecm/core/storage/dbs/DBSTransactionState.java

```diff
@@ -309,20 +334,44 @@ public boolean hasChild(String parentId, String name) {
         return docStates;
     }
 
-    public List<String> getChildrenIds(String parentId) {
+    public List<String> getAllChildrenIds(String parentId) {
+        return getChildrenIds(parentId, false, false);
+    }
+
+    public List<String> getChildrenIds(String parentId, boolean excludeSpecialChildren,
+            boolean excludeRegularChildren) {
         List<String> children = new ArrayList<>();
         Set<String> seen = new HashSet<>();
+        Set<String> specialChildrenTypes = Framework.getService(SchemaManager.class).getSpecialDocumentTypes();
+        boolean excludeChildren = excludeSpecialChildren || excludeRegularChildren;
         // check transient state
         for (DBSDocumentState docState : transientStates.values()) {
             String id = docState.getId();
             if (!parentId.equals(docState.getParentId())) {
                 continue;
             }
+            if (excludeChildren) {
+                boolean specialChild = specialChildrenTypes.contains(docState.getPrimaryType());
+                if (excludeSpecialChildren && specialChild || excludeRegularChildren && !specialChild) {
+                    continue;
+                }
+            }
             seen.add(id);
             children.add(id);
         }
-        // fetch from repository
-        List<State> states = repository.queryKeyValue(KEY_PARENT_ID, parentId, seen);
+        List<State> states;
+        if (!excludeChildren) {
+            states = repository.queryKeyValue(KEY_PARENT_ID, parentId, seen);
+        } else {
+            // fetch from repository depending on the filter flags
+            DBSQueryOperator operator = excludeSpecialChildren ? DBSQueryOperator.NOT_IN : DBSQueryOperator.IN;
+            states = repository.queryKeyValueWithOperator(KEY_PARENT_ID, parentId, KEY_PRIMARY_TYPE, operator,
+                    specialChildrenTypes, seen);
+        }
+        return internalGetChildren(children, states);
+    }
+
+    protected List<String> internalGetChildren(List<String> children, List<State> states) {
```

Comment:
This doesn't need to be a separate method anymore now that we're back to just one caller.